### DISCUSSION TITLE
withoutTrashed function not available on BelongsToMany relation

### DISCUSSION THREAD

**Discussion by gepopp** [laravel/nova-issues#4092] [Q&A]
- Laravel Version: 9.1
- Nova Version: 4.2.5
- PHP Version: 8.1

### Description:
withoutTrashed() not available on BelongsToMany relation, how can i get rid of the checkbox?
´

**Comment by crynobone** [comment]
If you want to submit a bug report please fill in the bug report with detail description as required via <https://github.com/laravel/nova-issues/issues/new?assignees=&labels=&template=1_Bug_report.md>. This feels like a question.

**Reply by gepopp** [reply]
What more description do u need than that a function available with other relationship fields is not available with that one?

**Reply by crynobone** [reply]
> withoutTrashed() not available on BelongsToMany relation

Yes, it not available because Eloquent with BelongsToMany relationship doesn't support soft deletes.

> how can i get rid of the checkbox?

Which checkbox? You still can add and run actions, and you can detach relations using the available checkbox. 

Or are you talking about something else? It's not clear here.

**Comment by gepopp** [comment]
I have a manyToMany relation between users and speeches with a pivot table with the columns:

- id
- speech_id
- user_id
- timestamps

the pivot model looks like:


namespace App\Models;

use Illuminate\Database\Eloquent\Relations\Pivot;


```

class SpeachUser extends Pivot {


    protected $guarded = [];


    public function speech() {

        return $this->belongsTo( Speach::class );
    }

    public function user() {

        return $this->belongsTo( User::class );
    }

}

```

Relations in the models are declared like:
User (supports soft deletes):

```
 public function speeches() {

        return $this->belongsToMany( Speech::class )
            ->withTimestamps()
            ->using(SpeechUser::class);
    }
```

the speech model (supports soft delets):

```
 public function speakers() {

        return $this->belongsToMany( User::class )
            ->withTimestamps()
            ->using(SpeechUser::class);
    }

```

in the nova resources the fields are declared as follows:
User:
 ```
BelongsToMany::make( 'Speeches', 'speeches', 'App\Nova\Speech' )
                         ->searchable(),

```

Speech:

```

            BelongsToMany::make('Speaker', 'speakers', 'App\Nova\User')
            ->searchable()
```

if i now go to user and click on Attach Speech it shows the "with trashed" chackbox beneath the search field:
![image](https://user-images.githubusercontent.com/2655976/164990976-50975acb-fea6-48d1-859a-02f40c94177e.png)

![image](https://user-images.githubusercontent.com/2655976/164990753-eb8ed4e7-a638-4646-b5bd-5ee609157a2d.png)

same thing the other way arround when i go to a speech and attach speaker,

if i try to add the withoutTrashed() to the BelongsToMany i get the error

`Method Laravel\\Nova\\Fields\\BelongsToMany::withoutTrashed does not exist.`

**Comment by gldrenthe89** [comment]
Why is this still a thing. I want that 'With Trashed' button gone. Why do i have to build a custom field to achieve this?

**Comment by Norgul** [comment]
I am using Nova 4 and this is still not available.
This is a **must have** feature. Enabling my users to pick from a models which are already soft deleted is a disaster!

Are there any news with regards to it?

**Comment by jeremynikolic** [comment]
After some digging, there has indeed been some misunderstanding of the issue/request at hand.
But all good now, sorry about that 🙏 

Indeed the BelongsToMany does not offer control over that "With trashed" checkbox visibility when editing the attached models. The current functionality "just" looks at if the attachable model class makes use of SoftDeletes, if it does then the checkbox appears else it does not.

We'll look at bringing this to Nova 5, coming soon 👍

**Reply by Norgul** [reply]
@jeremynikolic One remark though, I would consider "With trashed" option to not be a default, and instead of excluding it with `withoutTrashed()` I would assume that it was excluded by default and that you need to include it with `withTrashed()` instead. 

I mean, how many times do you really want a customer to make a relation to a deleted model?

**Reply by jeremynikolic** [reply]
That's a pretty valid point 🤔
Let me get back to you on this

**Reply by jeremynikolic** [reply]
Got reminded why this currently is the case 💡 

It's because this is slightly coupled with showing the "current" relation. So if we remove that checkbox, we would have issues  retrieving the current related model if it is soft deleted.

This is something we would need to look a bit deeper into and do some testing to see if we can de-couple. 
If you'd be so kind as to submit this to it's own discussion so we have it separated from this one 🙏

**Comment by vokamut** [comment]
I just hid it:
```css
[dusk="users-with-trashed-checkbox"] {
    display: none;
}
```

Where `users` is the name of the relation

**Reply by Norgul** [reply]
@vokamut where did you put that code?

**Reply by vokamut** [reply]
Put into the file `public/nova.css`.

And into the `NovaServiceProvider`:

```php
<?php

namespace App\Providers;

use Laravel\Nova\Nova;
use Laravel\Nova\NovaApplicationServiceProvider;

class NovaServiceProvider extends NovaApplicationServiceProvider
{
    public function boot(): void
    {
        parent::boot();

        Nova::style('nova', public_path('nova.css')); // <--------------------------
    }
}
```

**Reply by Norgul** [reply]
Would that hide all of them though? Not just a specific one?

**Reply by jeremynikolic** [reply]
The dusk identifier is being prefixed with the fields attribute. It should be specific enough 🤔

Pattern is as follow `{Attribute}-with-trashed-checkbox`

Not bulletproof but good workaround for now 👍

**Reply by vokamut** [reply]
Hide everything with `users` relation

**Comment by crynobone** [comment]
Released with Laravel Nova 5.0.0

Feel free to open up a new issue if you're still experiencing this problem on the latest version.
